src/main.rs

```diff
@@ -1,3 +1,829 @@
+
+fn cmd_forbid(
+    _out: &mut dyn Write,
+    _cli: &Cli,
+    _config: &Configs,
+    _metadata: &Metadata,
+    _sub_args: &ForbidArgs,
+) -> Result<(), Box<dyn Error>> {
+    // Forbid a crate's source for some version
+    unimplemented!()
+}
+
+fn cmd_audits(
+    _out: &mut dyn Write,
+    _cli: &Cli,
+    _config: &Configs,
+    _metadata: &Metadata,
+    _sub_args: &AuditsArgs,
+) -> Result<(), Box<dyn Error>> {
+    // ??? list audits? update audits?
+    unimplemented!()
+}
+fn cmd_suggest(
+    _out: &mut dyn Write,
+    _cli: &Cli,
+    _config: &Configs,
+    _metadata: &Metadata,
+    _sub_args: &SuggestArgs,
+) -> Result<(), Box<dyn Error>> {
+    // Suggest low-hanging-fruit reviews
+    unimplemented!()
+}
+fn cmd_diff(
+    _out: &mut dyn Write,
+    _cli: &Cli,
+    _config: &Configs,
+    _metadata: &Metadata,
+    _sub_args: &DiffArgs,
+) -> Result<(), Box<dyn Error>> {
+    // ??? diff something
+    unimplemented!()
+}
+
+fn cmd_vet(
+    out: &mut dyn Write,
+    cli: &Cli,
+    config: &Configs,
+    metadata: &Metadata,
+) -> Result<(), Box<dyn Error>> {
+    // Run the checker to validate that the current set of deps is covered by the current cargo vet store
+    trace!("vetting...");
+
+    let store_path = config.store_path();
+    let audit_inputs = config.audits();
+
+    let audited_path = store_path.join(AUDITED_TOML);
+    let trusted_path = store_path.join(TRUSTED_TOML);
+    let unaudited_path = store_path.join(UNAUDITED_TOML);
+
+    let _audited = Audited::new(load_deps_toml(&audited_path)?);
+    let unaudited = Unaudited::new(load_deps_toml(&unaudited_path)?);
+    let _trusted = Trusted::new(load_deps_toml(&trusted_path)?);
+
+    // Update audits (trusted.toml)
+    if !cli.locked && !audit_inputs.is_empty() {
+        unimplemented!("fetching audits not yet implemented!");
+    }
+
+    // TODO: do proper resolution on the 3 inputs.
+    //
+    // Relevant entries:
+    // * unaudited.version: this entire version is unaudited, but implicitly trusted
+    // * audited.forbid: this version is bad, do not use
+    // * audited.version: this entire version has been reviewed
+    // * audited.delta(x -> y): the changes from x to y 
+    // * trusted.*: I think the same as audited.* but separate for logistics
+    //   probably audited "overrides" trusted if they disagree? (via forbid?)
+    //
+    // If not for deltas, resolving packages would be fairly trivial.
+    // *With* deltas I think we want to have some DAG-like analysis where you start
+    // at the current version and look for a 'delta.to' that matches that version,
+    // and then recursively check `delta.from'
+    //
+    // One thing that's unclear is what should happen if 'delta' hops *over* a version
+    // i.e. if we have `audited.version = 5` and `audited.delta = 3 -> 7', does that
+    // allow us to accept version 7? This is kind of an incoherent state but it seems
+    // plausible with 'trusted' inputs imported from elsewhere!
```

Comment:
Loose notes on my initial reading of the format and how to "implement" it. Again needs to be rejigged based on book.

Last paragraph covers an "interesting question" about deltas that is probably still relevant.

Reply:
Regarding this last question: we should require an additional audited.delta = 5->3 or 5->7. Otherwise, someone could introduce something malicious in version 2, remove it in version 4, and add it back again in version 6.

Reply:
I agree that conceptually it can happen, but in theory "add it back in 6" would have been visible in the claimed delta of `3 -> 7`, no?

...OHH. right right right. I kept conceptually thinking of this as reviewing all the patches/PRs for that delta, but of course the expected/reasonable workflow is to diff the entire repo, and not see the patches themselves. Gross! Nice!